SetupDataPkg/ConfProfileMgrDxe/ConfProfileMgrDxe.c

```diff
@@ -0,0 +1,248 @@
+
+/**
+  Validate the chosen profile values against variable storage. If any value mismatch, write profile values to flash
+  and reset the system.
+
+  @param[in] ActiveProfileGuid  File guid for the currently active profile
+**/
+STATIC
+VOID
+ValidateActiveProfile (
+  )
+{
+  EFI_STATUS             Status;
+  CONFIG_VAR_LIST_ENTRY  *VarList     = NULL;
+  UINTN                  VarListCount = 0;
+  UINT32                 i;
+  BOOLEAN                ValidationFailure = FALSE;
+  UINTN                  Size;
+  UINT32                 Attributes;
+  VOID                   *Data = NULL;
+
+  Status = RetrieveActiveConfigVarList (&VarList, &VarListCount);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a - Retrieving Variable List failed - %r\n", __FUNCTION__, Status));
+    ASSERT (FALSE);
+    goto Done;
+  }
+
+  if ((VarList == NULL) || (VarListCount == 0)) {
+    // This should not be the case
+    DEBUG ((DEBUG_ERROR, "%a Retrieved config data is NULL.\n", __FUNCTION__));
+    ASSERT (FALSE);
+    goto Done;
+  }
+
+  for (i = 0; i < VarListCount; i++) {
+    Data = AllocatePool (VarList[i].DataSize);
+
+    if (Data == NULL) {
+      DEBUG ((DEBUG_ERROR, "%a Failed to allocate memory, size: %u var: %s.\n", __FUNCTION__, VarList[i].DataSize, VarList[i].Name));
+      ASSERT (FALSE);
+      continue;
+    }
+
+    Size = VarList[i].DataSize;
+
+    // Compare profile value against flash value
+    Status = gRT->GetVariable (
+                    VarList[i].Name,
+                    &VarList[i].Guid,
+                    &Attributes,
+                    &Size,
+                    Data
+                    );
+
+    // We should not fail to read any of these variables, if so, write variables, reset
+    // It is possible that we will fail to read the variable due to the size being different, indicating
+    // the variable has changed. That may be expected and should be handled as the case of the active profile
+    // not matching the variable store
+    if ((Status != EFI_BUFFER_TOO_SMALL) && EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "%a failed to read variable %s Status: (%r)!\n", __FUNCTION__, VarList[i].Name, Status));
+      ASSERT (FALSE);
+      ValidationFailure = TRUE;
+      goto Done;
+    }
+
+    if ((Size != VarList[i].DataSize) ||
+        (0 != CompareMem (Data, VarList[i].Data, Size)) ||
+        (Attributes != VarList[i].Attributes))
+    {
+      DEBUG ((DEBUG_ERROR, "%a active profile does not match variable store %s!\n", __FUNCTION__, VarList[i].Name));
+      ValidationFailure = TRUE;
+      goto Done;
+    }
+
+    FreePool (Data);
+    Data = NULL;
+  }
+
+Done:
+  if (Data != NULL) {
+    FreePool (Data);
+  }
+
+  if (ValidationFailure) {
+    // Write profile values and reset the system
+    for (i = 0; i < VarListCount; i++) {
+      Status = gRT->SetVariable (
+                      VarList[i].Name,
+                      &VarList[i].Guid,
+                      VarList[i].Attributes,

```

Comment:
Note that if the existing variable has a different attribute than what you are trying to set here, the SetVariable will fail. So you might want to delete it first.

Reply:
Will do.